graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+#![allow(dead_code, unused_variables)]
+use ast::*;
+use lexer::Punctuator::*;
+use lexer::Token::{
+  Eof,
+  Punctuator,
+  Name as TokenName,
+  IntValue as TokenInt,
+  FloatValue as TokenFloat,
+  StringValue as TokenString,
+};
+use lexer::{Lexer, Token};
+use source::Source;
+use std::iter::Peekable;
+
+pub type Res<'a, T> = Result<T, &'a str>;
+pub type DRes<'a, T> = Res<'a, (T, usize)>;
+
+pub struct Parser<'a> {
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
+    Parser {
+      lexer: Lexer::new(src.body).peekable(),
+      source: src,
+    }
+  }
+
+  fn loc(&self, lo: usize, hi: usize) -> Location<'a> {
+    Location {
+      start: lo,
+      end: hi,
+      source: self.source,
+    }
+  }
+}
+
+impl<'a> Iterator for Parser<'a> {
+  type Item = Res<'a, Definition<'a>>;
+  fn next(&mut self) -> Option<Res<'a, Definition<'a>>> {
+    if self.lexer.peek().is_some() {
+      Some(parse_definition(self))
+    } else {
+      None
+    }
+  }
+}
+
+macro_rules! on_do {
+ ($parser: ident, $($p: pat => $b: block)+) => {
```

Comment:
Huh? Doc this? Does this even need to be a macro? This seems to gain nothing but confusion by being a macro.